Added a Pull Request Template (#17856)
django/django: .github/pull_request_template.md

# Trac ticket number

> <!-- Replace [number] with the corresponding Track ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

Reviewer: ```suggestion
<!-- Replace [number] with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->
```

Reviewer 2: Great catch @LilyFoote, thank you! Fixing.

# Description
Briefly describe the problem or reason for the changes implemented in the PR

# Trac ticket

> Link to the accepted Trac ticket for this PR. All PRs except docs or tests changes need an accepted ticket. New features need community consensus.

Reviewer: I think a new feature ticket would need to already have community consensus to be an accepted ticket, so that bit seems a bit redundant to me.

Reviewer 3: _"All PRs except docs or tests changes need an accepted ticket."_

This sentence is not true and really misleading, check out [docs](https://docs.djangoproject.com/en/5.0/internals/contributing/writing-code/submitting-patches/#typo-fixes-and-trivial-documentation-changes). For example, small cleanups in the source code may not require a ticket, and new pages in docs may required a ticket.

# Tests

> Briefly describe any tests you have added or updated to test the changes introduced by the PR

Reviewer 4: Is that generally expected? I feel like having the tests in the checklist is good enough – asking people to describe the tests they've written feels like a sure-fire way to get the paragraph deleted entirely.

Reviewer 5: I agree 👍

Additionally when we're also draining people's cognitive resources by adding too much stuff to read so repetitive things should be removed.

Less is more :)

> # Tests
> Briefly describe any tests you have added or updated to test the changes introduced by the PR

Reviewer 2: We can repurpose this item to say that UI changes need screenshots:
```suggestion
# UI Changes
Please attach screenshots in light and dark mode for UI changes.
```

# Checklist

> - [ ] I have updated the Trac ticket’s flags

Reviewer 4: As Trac isn't super intuitive, I think it would be helpful to tell users what to do exactly (ie set "has patch" on the ticket).

Reviewer 2: ```suggestion
- [ ] I have updated the "has patch" Trac ticket
```

> - [ ] My commit message is written in past tense, mentions the ticket #, and ends with a dot

Reviewer 2: ```suggestion
- [ ] The commit message is written in past tense, mentions the ticket #, and ends with a dot
```

> - [ ] My PR targets the `main` branch. _Backport to other stable branches will be done by mergers when necessary._

Reviewer 2: ```suggestion
- [ ] This PR targets the `main` branch. (Backport to other stable branches will be done by mergers when necessary.)
```

> - [ ] I have added or updated tests to test the changes I made

Reviewer 2: ```suggestion
- [ ] I have added or updated tests relevant to the changes being proposed
```

> - [ ] I have added or updated relevant documentation (if applicable)

Reviewer 2: ```suggestion
- [ ] I have added or updated the docs, including release notes (if applicable)
```

> # More details

Reviewer 4: I'm not sure this section is a good idea, tbh, for multiple reasons: 

- When you open a PR, GitHub shows you a sidebar linking to the contributing.rst, which in turn links [to the contribution docs](https://docs.djangoproject.com/en/dev/internals/contributing/).
- The "Working with git" documentation is probably not the most relevant once people are already about to open a PR
- The PR template is opened in an editable text box, so using Markdown format is cumbersome, as you either have to manually select the link, or switch to the preview panel to click it
- IME, the template should be as short as possible (so people don't just delete it entirely), and only contain things that should end up in the PR text – not least of all because you can search PRs, and leaving this section in would potentially clutter up the search.

Reviewer 2: > I'm not sure this section is a good idea, tbh, for multiple reasons:
> 
>     * When you open a PR, GitHub shows you a sidebar linking to the contributing.rst, which in turn links [to the contribution docs](https://docs.djangoproject.com/en/dev/internals/contributing/).

While this link is helpful, I think it too broad and unspecific about the stage the author is at this point.

>     * The "Working with git" documentation is probably not the most relevant once people are already about to open a PR

Good point about the git docs, though there are a few sections at the bottom of that page that are very relevant for when proposing a PR and then following up during reviews. It describes rebasing and other relevant processes. I'm proposing that we tailor the link to these sections instead.

>     * The PR template is opened in an editable text box, so using Markdown format is cumbersome, as you either have to manually select the link, or switch to the preview panel to click it
> 
>     * IME, the template should be as short as possible (so people don't just delete it entirely), and only contain things that should end up in the PR text – not least of all because you can search PRs, and leaving this section in would potentially clutter up the search.

We have many contributors not following the process of having a ticket, adding tests, adding docs, setting the trac flags for a PR. I believe that they are not following the expected procedure because they don't know about it (not because they ignore it). And I think that they don't know about it because they got lost in the contributing docs (I've been there!). More so, I think that while is likely that (new?) contributors are familiar with GH, they are not with other systems or may be discouraged by the need to jump between the docs site, the trac site, and this site.

With this information, do you think there is something we could do to improve the contributor experience and, consequently, help reviewers receive more complete PRs?

Author: > The PR template is opened in an editable text box, so using Markdown format is cumbersome, as you either have to manually select the link, or switch to the preview panel to click it

I think I agree with @rixx in that this information might be better placed elsewhere, since having doc links in a PR template can make the template overcrowded with info, and the fact that even opening the links from the template is a bit cumbersome is a good point.

I agree that it's easy to get lost in the contributing docs, but maybe that's something we need to fix separately?

For more information please refer to the following links:

> - [Submitting Patches documentation](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/)

Reviewer 2: ```suggestion
- [Pull Request Checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist)
```

> - [Working with Git documentation](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/)

Reviewer 2: ```suggestion
- [Working with GitHub](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#publishing-work)
```

# Description

> Briefly describe the problem or reason for the changes implemented in the PR

Reviewer 6: ```suggestion
Briefly describe the problem or reason for the changes implemented in the PR.
```

> # Trac ticket
> Link to the Trac ticket for this PR (all PRs except really trivial issues or 
> tests adding need an accepted ticket).

Reviewer 7: I think we should mention that you can write `ticket-<number>` and it’s [autolinked](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/configuring-autolinks-to-reference-external-resources).

```suggestion
# Trac ticket

<!-- Replace <number> with ticket number, or delete the line and write N/A if
this is a trivial PR. -->

ticket-<number>
```

Reviewer 2: TIL! :heart_eyes:

# UI Changes
Please attach screenshots in light and dark mode for UI changes.

> 

Reviewer 7: Let’s put this as a final point in the checklist. It doesn’t need a separate section.

```
- [ ] For UI changes, I have attached screenshots in both light and dark modes.
```

# Checklist

> - [ ] I have updated the `has patch` flag in the Trac ticket

Reviewer 7: Use quotes rather than code, and say “checked” for clarity. Also turn into a sentence.
```suggestion
- [ ] I have checked the “has patch” flag in the Trac ticket.
```

- [ ] The commit message is written in past tense, mentions the ticket #, and

>  ends with a dot

Reviewer 7: It’s [“period” in American English](https://en.wikipedia.org/wiki/Full_stop), “dot” would only refer to use as an operator in code, like `self.cheese`. Also add one here.
```suggestion
 ends with a period.
```

- [ ] This PR targets the `main` branch. (Backport to other stable branches 
will be done by mergers when necessary.)

> - [ ] I have added or updated tests relevant to the changes being proposed
> - [ ] I have added or updated the docs, including release notes (if applicable)

Reviewer 7: Turn into sentences, edit for clarity
```suggestion
- [ ] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
```
